Update toolchain format and convenience scripts

b-n opened the pull request:
This could probably be done without the `toml` ruby gem and using some regex's, but this felt safer.

Some proof that this works: https://github.com/b-n/awt/commit/5b17a001da485ad449fefde4de144b7d5ec58998 and https://github.com/b-n/awt/pulls?q=is%3Apr+author%3Aapp%2Fdependabot+ :smile: 

Example diff after changing `rust-toolchain.toml` and running `rake toolchain:sync`

```diff
diff --git a/.github/workflows/ci.yaml b/.github/workflows/ci.yaml
index f69d2120e0..ff9a41069c 100644
--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -29,7 +29,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/build-and-test@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Compile
         run: cargo build --workspace --verbose
@@ -53,7 +53,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/lint-and-format@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Check artichoke formatting
         run: cargo +nightly fmt --check
@@ -86,7 +86,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/lint-and-format@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Check fuzz formatting
         run: cargo +nightly fmt --check
@@ -109,7 +109,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/lint-and-format@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Check spec-runner formatting
         run: cargo +nightly fmt --check
@@ -140,7 +140,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/lint-and-format@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Check ui-tests formatting
         run: cargo +nightly fmt --check
@@ -178,7 +178,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/build-and-test@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Check artichoke-load-path with no default features
         run: cargo check --verbose --no-default-features --all-targets --profile=test
@@ -385,7 +385,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/check-minimal-versions@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Generate minimal versions lockfile in root workspace
         run: cargo +nightly generate-lockfile -Z minimal-versions
diff --git a/.github/workflows/crash.yaml b/.github/workflows/crash.yaml
index 41b45bd626..450a737dd2 100644
--- a/.github/workflows/crash.yaml
+++ b/.github/workflows/crash.yaml
@@ -20,7 +20,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/build-and-test@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Compile debug
         run: cargo build --verbose --bin spec-runner
diff --git a/.github/workflows/spec-state.yaml b/.github/workflows/spec-state.yaml
index bf3b566f7d..5fdd786c0f 100644
--- a/.github/workflows/spec-state.yaml
+++ b/.github/workflows/spec-state.yaml
@@ -24,7 +24,7 @@ jobs:
       - name: Install Rust toolchain
         uses: artichoke/setup-rust/build-and-test@v1.9.0
         with:
-          toolchain: "1.70.0"
+          toolchain: "1.71.0"
 
       - name: Compile spec-runner
         run: cargo build --release --verbose --bin spec-runner
diff --git a/Cargo.toml b/Cargo.toml
index b6ec57666b..26e9163570 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -62,7 +62,7 @@ members = [
 
 [workspace.package]
 edition = "2021"
-rust-version = "1.70.0"
+rust-version = "1.71.0"
 license = "MIT"
 repository = "https://github.com/artichoke/artichoke"
 homepage = "https://github.com/artichoke/artichoke"
diff --git a/fuzz/Cargo.toml b/fuzz/Cargo.toml
index 65698b3756..33116295f8 100644
--- a/fuzz/Cargo.toml
+++ b/fuzz/Cargo.toml
@@ -4,7 +4,7 @@ version = "0.0.0"
 authors = ["Automatically generated"]
 publish = false
 edition = "2021"
-rust-version = "1.70.0"
+rust-version = "1.71.0"
 license = "MIT"
 
 [package.metadata]
diff --git a/rust-toolchain.toml b/rust-toolchain.toml
index 22048ac5bb..aa464261d8 100644
--- a/rust-toolchain.toml
+++ b/rust-toolchain.toml
@@ -1,2 +1,2 @@
 [toolchain]
-channel = "1.70.0"
+channel = "1.71.0"
diff --git a/spec-runner/Cargo.toml b/spec-runner/Cargo.toml
index 09be8912a2..26262104b5 100644
--- a/spec-runner/Cargo.toml
+++ b/spec-runner/Cargo.toml
@@ -2,7 +2,7 @@
 name = "spec-runner"
 version = "0.7.0"
 edition = "2021"
-rust-version = "1.70.0"
+rust-version = "1.71.0"
 license = "MIT"
 authors = ["Ryan Lopopolo <rjl@hyperbo.la>"]
 description = "Binary for running Ruby Specs with Artichoke Ruby"
diff --git a/ui-tests/Cargo.toml b/ui-tests/Cargo.toml
index fe9f33c656..1d0319ab90 100644
--- a/ui-tests/Cargo.toml
+++ b/ui-tests/Cargo.toml
@@ -4,7 +4,7 @@ version = "0.0.0"
 authors = ["Automatically generated"]
 publish = false
 edition = "2021"
-rust-version = "1.70.0"
+rust-version = "1.71.0"
 license = "MIT"
 
 [dependencies]
```

Fixes: #2590

b-n committed:
Move toolchain to .toml and format respectively

b-n committed:
Update `rake toolchain:sync` to read from the TOML verison

b-n commented on Rakefile:
```diff
@@ -196,22 +197,22 @@ namespace :toolchain do
 
         if (existing_version = contents.match(regexp))
-          File.write(file, contents.gsub(regexp, next_rust_version)) if existing_version != rust_version
+          File.write(file, contents.gsub(regexp, next_rust_version)) if existing_version != next_rust_version
```
`existing_version` contains the full regexp string (e.g. `rust-version = "1.70.0"`), so we should  match against the full `next_rust_version`, rather than just the toolchain version.

lopopolo replied:
Yup this is intentional, just in case we ever have a crate version that has the same version string as the Rust toolchain version.

b-n commented:
~Converting to draft whilst I go and update the docs that I found referencing this file~

I've made the most relevant changes. There are still referencs to `rust-toolchain` but without a .toml extension since I feel like these sentences still make sense in context without specifying the full path.

b-n committed:
Update docs refering to the non .toml rust-toolchain

lopopolo commented on BUILD.md:
````diff
@@ -52,14 +52,8 @@ prompts: <https://rustup.rs/>. This requires a download of Visual Studio (the
 selected through the VS component installer. (I'm not sure which packages are
 required; I selected them all.)
 
-Once you have rustup, you can install the Rust toolchain needed to compile
-Artichoke.
-
-```sh
-rustup toolchain install "$(cat rust-toolchain)"
-rustup component add rustfmt
-rustup component add clippy
-```
````
Can we rework this section to have folks manually install `rustfmt` and `clippy`? I also think the rustup command can be condensed to this:

```console
$ rustup component add rustfmt clippy
info: component 'rustfmt' for target 'x86_64-apple-darwin' is up to date
info: component 'clippy' for target 'x86_64-apple-darwin' is up to date
```

lopopolo commented on rust-toolchain.toml:
```diff
@@ -0,0 +1,3 @@
+[toolchain]
+channel = "1.70.0"
+components = ["clippy", "rustfmt"]
```
Can we remove these components from the toolchain file?

I previously asked the `rustup` team about this and got guidance that the toolchain file shouldn't be used for developer convenience. I'm also not sure how defining the components here interacts with `artichoke/setup-rust` action in CI.

See:

- https://github.com/rust-lang/rustup/issues/2686#issuecomment-1083831334
- https://github.com/rust-lang/rustup/issues/2686#issuecomment-1084179901

lopopolo commented on Rakefile:
```diff
@@ -196,22 +197,22 @@ namespace :toolchain do
   task sync: %i[sync:manifests sync:ci]
 
-  rust_toolchain = File.read('rust-toolchain').chomp
+  rust_toolchain = TOML.load_file('rust-toolchain.toml')
+  toolchain_version = rust_toolchain['toolchain']['channel']
```
this is great!

b-n replied:
I did some ol `symbolize_keys` in tomlrb. Thoughts?

lopopolo replied:
sweet, I like it.

lopopolo commented on Gemfile:
```diff
@@ -6,3 +6,4 @@ gem 'bundler-audit', '~> 0.9', require: false
 gem 'rake', '>= 12.3.3', require: false
 gem 'rubocop', '~> 1.51', require: false
 gem 'rubocop-rake', '~> 0.6', require: false
+gem 'toml', '~> 0.3', require: false
```
This dep is:

- https://github.com/jm/toml
- https://rubygems.org/gems/toml/versions/0.3.0

Plenty of downloads on RubyGems with 18M+. Not sure how maintained it is: the last commit was in December 2021.

b-n replied:
Ah, you know what, I've found a better one (in all regards): https://github.com/fbernier/tomlrb

Better in that:
- No additional dependencies
- It's had more love over the years
- It explicitly supports TOML v1.0.0
- It's benchmarked itself against the only other compliant TOML library (toml-rb)
- [depending on whether you like it] it also has symbolize_keys option
- Even more downloads than `toml` at 50mil

Also, it's MIT (same as `toml`, but just calling it out)

lopopolo replied:
Awesome. Change in the playground PR is at artichoke/playground@0070a9aecce16a581dc2090737bf7ae868b17fbc.

lopopolo commented on Rakefile:
```diff
@@ -4,6 +4,7 @@ require 'open-uri'
 require 'shellwords'
 require 'bundler/audit/task'
```
Since you're mucking around here, can we add a blank line to separate official Ruby deps from third party deps?

```suggestion
require 'shellwords'

require 'bundler/audit/task'
```

lopopolo commented:
Thanks for taking this @b-n 🙇 Just a couple of tweaks and this should be good to go.

lopopolo commented:
I made a similar change to the playground modeled off of this PR:

- https://github.com/artichoke/playground/pull/1066

b-n committed:
Feedback from review

BUILD.md changes - I can see why rustup people don't say you should add
it, because in theory a human can do what these tools do, so they're not
entirely mandatory.

b-n committed:
Use a more maintained and up to date TOML parser

lopopolo reviewed (approved):
looks good

lopopolo commented:
This broke downstream builds in the nightly and docker nightly repos. Fixes are here:

- https://github.com/artichoke/nightly/pull/172
- https://github.com/artichoke/docker-artichoke-nightly/pull/164

I'll merge now to get the builders unstuck. Feel free to do a post-merge code review if you'd like. I used a trick of setting `python` as the shell for a GitHub Actions step to use the built-in `tomllib` to parse the toolchain file.
